src/linker/Linker/Annotations.cs

```diff
@@ -552,6 +553,31 @@ public bool SetPreservedStaticCtor (TypeDefinition type)
 			return marked_types_with_cctor.Add (type);
 		}
 
+		public bool HasRequiresUnreferencedCodeOnTypeHierarchy (TypeDefinition type)
```

Comment:
Why do we need to collect this on the type hierarchy?

I would expect:

```csharp
[RUC]
class Foo
{
}

class Bar : Foo
{
}
```

To just give me a warning when we're analyzing Bar's constructor (it's going to call into Foo's constructor and that one is RUC).

It's also how the C# compiler handles the `[Obsolete]` attribute. It doesn't auto propagate to derived classes.

The cache costs extra complexity and memory and I'm not sure it's giving us anything useful.

Reply:
That's true, maybe I should have used other name for the functions and cache, but the current changes don't auto propagate between derived classes the cache is only going over nested types. For the scenario you mentioned there will only be a warning in `class Bar : Foo` because it calls Foo constructor just like [Obsolete] and is also one of the tests